Libsecret support

alex3kov opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [X] New feature

---

### Description of your *pull request* and other information

Add functionality for retrieving and storing credentials for domain/username combinations - addresses #11642. Previous PR #10610 is a bit rudimentary and apparently abandoned by author.

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -11,6 +11,10 @@
 import random
 import sys
 
+import gi
```
This must be a soft dependency.

alex3kov replied:
How do you do this in this project - with `try:`/`except ImportError:`? I can't find anything like requirements files or `install_requires` in setup.py.

yan12125 replied:
Here's an example:
https://github.com/rg3/youtube-dl/blob/909191de9154bf289b333cfe01b8e88e3ac1fefc/youtube_dl/downloader/hls.py#L5-L9

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
     if opts.usetitle and opts.useid:
         parser.error('using title conflicts with using video ID')
-    if opts.username is not None and opts.password is None:
+    if opts.username is not None and opts.password is None and opts.password_from_keyring is True:
```
As already said in https://github.com/rg3/youtube-dl/pull/10610#issuecomment-246117414 you must not touch the current behavior. Similarly to `--netrc` it should not work together with `-u` and `-p` and print an error.

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
-    if opts.username is not None and opts.password is None:
+    if opts.username is not None and opts.password is None and opts.password_from_keyring is True:
+        # extract domain names, check if all videos are from the same domain.
+        all_domains = set(map(lambda url: compat_urllib_parse.urlparse(url).netloc, all_urls))
```
No. Extractors are identified by special keys not domains, see `_NETRC_MACHINE`.

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
+    if opts.username is not None and opts.password is None and opts.password_from_keyring is True:
+        # extract domain names, check if all videos are from the same domain.
+        all_domains = set(map(lambda url: compat_urllib_parse.urlparse(url).netloc, all_urls))
+        if len(all_domains) > 1:
```
No.

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
+        if len(all_domains) > 1:
+            parser.error('You passed URLs from more than one domain - supplying credentials from command line is not supported in this case.')
+        domain_name = all_domains.pop()
+        password = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"user-name": opts.username, "domain-name": domain_name}, None)
```
No. Passwords should be retrieved at the time of usage, see `_get_login_info`.

alex3kov replied:
This is troublesome, because AFAICS libsecret can only return a single value (password) - there's no way to retrieve other attributes of a keyring record. And `_get_login_info` needs to return both username and password.  
This ties in with your other comment https://github.com/rg3/youtube-dl/pull/14674#discussion_r148960973  
Basically, `--keyring` would be used as alternative to `--password`/interactive password. Would the following behaviour be acceptable? It seems to be working, could be refined further.  
1) Just `--username` = password prompt.
2) `--username` + `--keyring` = libsecret tries to look up password by special key (e.g. youtube) and username and sets `opts.password`. If password not found in keyring - throw error (or maybe fall back on password prompt to set `opts.password`).
3) `--username` + `--keyring` + `--password` = throw error  .

The only alternative I see atm is storing both username and password in a keyring record's secret with some separator, but that will lead to no end of parsing issues...

dstftw replied:
Well, some just cut corners like [this](https://github.com/tommie-lie/rhythmbox-pandora/blob/45fe8e755d78c04ddeb348bf09a377c94ec8832c/rhythmbox-pandora/PandoraAccount.py#L27). A better way is to store a formatted string (JSON/XML/whatever). However, it indeed looks like it's not designed for storing multiple values. That brings to something alike you propose apart from you should not set `opts.password` yourself. If no password is found in keyring an error should be raised since no interactive actions are allowed in embeddable part of the code.

alex3kov replied:
All the choices so far are a bit hacky:  
1) combination of `--username` and `--keyring` is just bad - can't handle videos from multiple domains passed to `youtube-dl`
2) Strings with `\n` separator are hard (if not impossible) to store in a keyring through `secret-tool`. It would be fine if we were storing that string from within `youtube-dl`, but @dstftw is against it.
3) XML has >2 characters that need escaping ([source](https://stackoverflow.com/questions/730133/invalid-characters-in-xml))
4) JSON has 2 characters that need escaping - `"` and `\` ([source](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf)), we can instruct users to do that in README.md. JSON module is included in Python 2.6, same as `youtube-dl`'s lowest Python requirement.  

JSON looks like the most passable. What do you think of https://github.com/rg3/youtube-dl/pull/14674/commits/3be4a1db6b6e144390cb337888daba6c4b52c852?

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
+            label = "Youtube-dl: " + domain_name + " password"
+            Secret.password_store_sync(LIBSECRET_SCHEMA, attributes, Secret.COLLECTION_DEFAULT, label, supplied_password, None)
+        password = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"user-name": opts.username, "domain-name": domain_name}, None)
+        opts.password = password
```
No. You should not shadow the original explicitly provided password.

alex3kov replied:
Removed - https://github.com/rg3/youtube-dl/commit/5b295effdd8ff980f319c975a9ec4b69879d5358

dstftw commented on youtube_dl/options.py:
```diff
@@ -366,6 +366,10 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
         dest='videopassword', metavar='PASSWORD',
         help='Video password (vimeo, smotri, youku)')
+    authentication.add_option(
+        '--password-from-keyring',
```
Should be `--gnome-keyring` or so for symmetry with `--netrc`.

alex3kov replied:
1) libsecret is supposed to be keyring-agnostic and work with either gnome-keyring or kwallet using DBus "Secret Service", so the name should be something more generic than `--gnome-keyring`.
2) I don't get your point - how is `--netrc` "symmetric" with `--gnome-keyring`?  

What option name would you suggest?

dstftw replied:
Option name should not contain `--password-from`. Select one that describes the mechanism in the most generic way. `--netrc` means the netrc mechanism will be used, symmetrically keyring option name should denote the keyring mechanism (or whatever it's properly called) will be used.

alex3kov replied:
Renamed to `--keyring`. Not 100% accurate, but seems more intuitive for users than `--libsecret`. https://github.com/rg3/youtube-dl/pull/14674/commits/ce0148424a15bda20fb46f67db5346a2ac836b36

dstftw commented on youtube_dl/__init__.py:
```diff
@@ -141,7 +147,22 @@ def _real_main(argv=None):
+            supplied_password = compat_getpass('Type account password and press [Return]: ')
+            attributes = {"user-name": opts.username, "domain-name": domain_name}
+            label = "Youtube-dl: " + domain_name + " password"
+            Secret.password_store_sync(LIBSECRET_SCHEMA, attributes, Secret.COLLECTION_DEFAULT, label, supplied_password, None)
```
No. youtube-dl should not store passwords.

alex3kov replied:
Removed - https://github.com/rg3/youtube-dl/pull/14674/commits/5b295effdd8ff980f319c975a9ec4b69879d5358

yan12125 commented on youtube_dl/extractor/common.py:
```diff
@@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)
+                if secret is None:
+                    raise netrc.NetrcParseError('Can\'t find credentials in keyring for "' + netrc_machine + '"')
```
Did you use the netrc format for data stored in libsecret backends? If not don't use ```NetrcParseError```.

alex3kov replied:
Replaced with `KeyError` (not sure if something else would be more appropriate) in https://github.com/rg3/youtube-dl/commit/3be4a1db6b6e144390cb337888daba6c4b52c852

yan12125 commented on youtube_dl/extractor/common.py:
```diff
@@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
+                print('JSON Parsed: ' + str(secret_json))
+                print('JSON User: ' + secret_json['u'])
+                print('JSON Password: ' + secret_json['p'])
+            except (IOError, netrc.NetrcParseError) as err:
```
ImportError and JSON parsing errors should be handled as well

alex3kov replied:
Added handling https://github.com/rg3/youtube-dl/commit/3be4a1db6b6e144390cb337888daba6c4b52c852

yan12125 commented on youtube_dl/extractor/common.py:
```diff
@@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)
```
What does "json" mean?

alex3kov replied:
Nothing - accidental github push.

yan12125 commented on youtube_dl/extractor/common.py:
```diff
@@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
+                secret_dict = json.loads(secret)
+                print('JSON Parsed: ' + str(secret_json))
+                print('JSON User: ' + secret_json['u'])
+                print('JSON Password: ' + secret_json['p'])
```
Better to use more meaningful names than u or p. That helps users to understand what youtube-dl are using from their secret storages.

alex3kov replied:
Replaced with user/pass in https://github.com/rg3/youtube-dl/commit/3be4a1db6b6e144390cb337888daba6c4b52c852

yan12125 commented on youtube_dl/__init__.py:
```diff
@@ -309,6 +313,7 @@ def parse_retries(retries):
 
     ydl_opts = {
         'usenetrc': opts.usenetrc,
+        'usekeyring': opts.usekeyring,
```
New options need to be documented in the docstring of ```YoutubeDL``` class.

alex3kov replied:
Documented in https://github.com/rg3/youtube-dl/pull/14674/commits/3be4a1db6b6e144390cb337888daba6c4b52c852

yan12125 commented on youtube_dl/extractor/common.py:
```diff
@@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
```
Better to wrap such long lines.

alex3kov replied:
Wrapped in https://github.com/rg3/youtube-dl/pull/14674/commits/3be4a1db6b6e144390cb337888daba6c4b52c852

alex3kov commented:
Sorry, I accidentally pushed one of intermediary drafts. Will post 
messages addressing all points once code gets up to speed.

/--Regards, Aleksei

/
On 10/11/17 18:37, Yen Chi Hsuan wrote:

*@yan12125* requested changes on this pull request.

Looks like this is still a work in progress? I leave some comments for 
current issues.

------------------------------------------------------------------------

In youtube_dl/extractor/common.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150261829>:

> @@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
              except (IOError, netrc.NetrcParseError) as err:
                  self._downloader.report_warning(
                      'parsing .netrc: %s' % error_to_compat_str(err))
+
+        if self._downloader.params.get('usekeyring', False):
+            try:
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)
+                if secret is None:
+                    raise netrc.NetrcParseError('Can\'t find credentials in keyring for "' + netrc_machine + '"')

Did you use the netrc format for data stored in libsecret backends? If 
not don't use |NetrcParseError|.

------------------------------------------------------------------------

In youtube_dl/extractor/common.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150262032>:

> +        
+        if self._downloader.params.get('usekeyring', False):
+            try:
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)
+                if secret is None:
+                    raise netrc.NetrcParseError('Can\'t find credentials in keyring for "' + netrc_machine + '"')
+                print(str(secret))
+                secret_dict = json.loads(secret)
+                print('JSON Parsed: ' + str(secret_json))
+                print('JSON User: ' + secret_json['u'])
+                print('JSON Password: ' + secret_json['p'])
+            except (IOError, netrc.NetrcParseError) as err:

ImportError and JSON parsing errors should be handled as well

------------------------------------------------------------------------

In youtube_dl/extractor/common.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150262249>:

> @@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
              except (IOError, netrc.NetrcParseError) as err:
                  self._downloader.report_warning(
                      'parsing .netrc: %s' % error_to_compat_str(err))
+
+        if self._downloader.params.get('usekeyring', False):
+            try:
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)

What does "json" mean?

------------------------------------------------------------------------

In youtube_dl/extractor/common.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150262538>:

> @@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
              except (IOError, netrc.NetrcParseError) as err:
                  self._downloader.report_warning(
                      'parsing .netrc: %s' % error_to_compat_str(err))
+
+        if self._downloader.params.get('usekeyring', False):
+            try:
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})
+                secret = Secret.password_lookup_sync(LIBSECRET_SCHEMA, {"extractor": "json", "app": "youtube-dl"}, None)
+                if secret is None:
+                    raise netrc.NetrcParseError('Can\'t find credentials in keyring for "' + netrc_machine + '"')
+                print(str(secret))
+                secret_dict = json.loads(secret)
+                print('JSON Parsed: ' + str(secret_json))
+                print('JSON User: ' + secret_json['u'])
+                print('JSON Password: ' + secret_json['p'])

Better to use more meaningful names than u or p. That helps users to 
understand what youtube-dl are using from their secret storages.

------------------------------------------------------------------------

In youtube_dl/__init__.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150262783>:

> @@ -309,6 +313,7 @@ def parse_retries(retries):
  
      ydl_opts = {
          'usenetrc': opts.usenetrc,
+        'usekeyring': opts.usekeyring,

New options need to be documented in the docstring of |YoutubeDL| class.

------------------------------------------------------------------------

In youtube_dl/extractor/common.py 
<https://github.com/rg3/youtube-dl/pull/14674#discussion_r150262917>:

> @@ -826,6 +826,24 @@ def _get_netrc_login_info(self, netrc_machine=None):
              except (IOError, netrc.NetrcParseError) as err:
                  self._downloader.report_warning(
                      'parsing .netrc: %s' % error_to_compat_str(err))
+
+        if self._downloader.params.get('usekeyring', False):
+            try:
+                import gi
+                gi.require_version('Secret', '1')
+                from gi.repository import Secret
+                LIBSECRET_SCHEMA = Secret.Schema.new("io.github.rg3.youtube-dl.Store",Secret.SchemaFlags.DONT_MATCH_NAME,{"extractor": Secret.SchemaAttributeType.STRING,"app": Secret.SchemaAttributeType.STRING})

Better to wrap such long lines.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub 
<https://github.com/rg3/youtube-dl/pull/14674#pullrequestreview-75788906>, 
or mute the thread 
<https://github.com/notifications/unsubscribe-auth/ADct8sJzMvt76B-HcL8pO9C8AzX60GiOks5s1G3RgaJpZM4QSaOJ>.
